src/sources/audiosource.cpp

```diff
@@ -217,10 +217,15 @@ bool AudioSource::verifyReadable() {
     return true;
 }
 
-WritableSampleFrames AudioSource::clampWritableSampleFrames(
+std::optional<WritableSampleFrames> AudioSource::clampWritableSampleFrames(
         WritableSampleFrames sampleFrames) const {
-    const auto readableFrameIndexRange =
+    const auto clampedFrameIndexRange =
             clampFrameIndexRange(sampleFrames.frameIndexRange());
+    if (!clampedFrameIndexRange) {
+        return std::nullopt;
```

Comment:
Here for example. 
Bothering the calling code with the null value is an extra layer of complexity. 
If the clampedFrameIndexRange is empty or invalid, in both cases the result is an empty buffer. 
The difference is only one frame offset.